### ISSUE TITLE
Support for F# record syntaxes for C# defined records

### ISSUE THREAD

**Issue by smoothdeveloper** [fsharp/fslang-suggestions#1138]
# Title of Suggestion

I propose we add support to F# language for
* what C# calls "non destructive mutation", cloning of records with arbitrary changes: `{ myRecord with ... }`
* record constructor syntax
* type inference on record field names

to C# defined record types.

## Pros

* better mapping of the same concepts, so they can be used idiomatically in respective languages (from F# side in this suggestion)
* less friction in multi-language codebases
* reinforces the "simple F#"

## Cons 

* adding logic in F# compiler, which is tied to knowledge of C# codegeneration idioms: need to keep track of potential changes that would impact this support as C# compiler evolves
* adding some complexity in the components currently supporting the record syntax bits in the type checker: longer compile times

## Extra information

Estimated cost: M

Related suggestions:

https://github.com/dotnet/fsharp/issues/13007#issuecomment-1115259153 (C# X support in F# Y issue)
https://github.com/dotnet/csharplang/discussions/6067 (co issue for C# compiler, discussions revolves around F# adjusting code generation for F# records to make it look like what C# compiler expects)

**Comment by vzarytovskii** [comment]
It sounds reasonable to me, and should be straightforward as long as the scope is only about the `with` and init.

It will, however, get complicated, if we want full compatibility, such as "converting"  C# records to F# ones and vice versa, we will need to deal with the inheritance of C# records then.

**Comment by smoothdeveloper** [comment]
@vzarytovskii, thanks!

I probably miss other features in the top bullet list that need to be considered, but could you expand on what you mean about "converting" as I am not sure which feature or aspect it relates to, I kind of connect it to the way anonymous records can be extended by adding properties, making a new type.

Is it this particular area of the language?

I am personally ok with limiting the feature scope to what is easiest to tackle, if it allows to cover what felt most needed in the initial list I've put.

This request is not about supporting inheritance of F# records, but I see that consuming the C# ones involves complexity related to records possibly forming a hierarchy.

**Comment by dsyme** [comment]
I'm in two minds about `{ r with X = 1 }` for C# records.  

Let's assume construction is `R(X=1, Y=2)`.  That seems reasonable.  Then in this case it just seems odd that we would switch to F# record syntax for copy-and-update.  Why not `r.With(X=1)` for example? 

If we did support `{ r with X = 1 }` then it would seem very likely people would expect `{ X = 1; Y = 2 }` as well.  

It's not a huge problem if only `{ r with X = 1 }` is supported but it's just a little awkard.

As an aside, another problems people note with `{ a with P=1 }` is that a generic record type can't change generic parameter, e.g. 
```fsharp
type R<'T> = { X : 'T; Id: string }
let r1 = { X = 1; Id = "a" } // R<int>
{r1 with X = "one" }  // doesn't become a R<string>
```
From the PL design perspective in theory this _could_ become a different type, but in F# it doesn't.   It does for anonymous records
```fsharp
let r1 = {| X = 1; Id = "a" |} // {| X: int; Id: string |}
{| r1 with X = "one" |}  // {| X: string; Id: string |}
```
However because of the way C# compiles `with` to a `<Clone>$` method there is no sceop to allow this for C#-interop `with`.

**Comment by dsyme** [comment]
There's another question whether C# records can be used as inputs to anonymous record syntax, e.g. 

`{| csharpRecord with Z = 1 |}`

F# records **can** be used as inputs to anonymous records.  So this should be possible for C# records too:

```fsharp
type R = { X : int; Id: string }
let r1 = { X = 1; Id = "a" } // R<int>
{| r1 with X = "one" |}
```

**Comment by smoothdeveloper** [comment]
@dsyme, a minor comment on

> The construction syntax for C# records is `R(X=1, Y=2)` and works today.

I think `R(1,2)` also works, and it is kind of not great. IMO, it should be giving a warning similar to https://github.com/fsharp/fslang-design/blob/main/drafts/FS-1095-requirenamedargumentattribute.md

Ideally, it should turn into record initialiser syntax proper, and other syntaxes should give warnings asking to use that syntax only.

Based on discussion on the C# co-thread (not gonna happen, unless F# code gen adjusts), the underlying code-gen for C# records would need to use the same `Clone` thing, I think this increases the complexity of what I envisioned before having those discussions.

All points you give and that were made on the C# co-thread are great insight of the subtleties into supporting this.

**Comment by dsyme** [comment]
Yes, we can generate `<Clone>$`, `init`, `reqd`, I don't see why not.

**Comment by smoothdeveloper** [comment]
> @vzarytovskii: I guess this is one the controversial things, whether we want F# record syntax for C# records.

> @dsyme My understanding is that these are positional parameters in C#, and so positional invocation is ok. Named properties in C# are R() { X = 1, Y = 2 }

To me, similar construct should flow simply, with niceties / idioms of the language they are consumed from, giving the "compiler is your friend" and "language is simple" feel.

True, they are positional in C#, but since F# doesn't allow usage of positional parameter for construction of F# defined records, it really should not encourage it for C# defined records, IMO.

Now, if I read @dsyme comments properly, it assumes that F# will distinguish C# records and expose `.With` object syntax for non destructive mutation, this remain idiomatic to C#, but I am unsure the distinction is worth it, if we see no issue with endorsing the same syntax we have for F# defined records.

> @vzarytovskii: Maybe I'm mistaken about what you mean here, but if i understood it correctly - i think generating Clone for F# records (to have them compatible with C# with syntax) is a reasonable thing to do.
> @dsyme: Yes, we can generate <Clone>$

This will make C# enable the `with` syntax on records defined in both languages, and also, expose `.With` object syntax to F# if we'd not go the (simpler to me) other route.

Overall, isn't it keeping F# simpler, if records still look like records, no matter which language it is defined in? with minor caveats for things that can be considered side cases (polymorphism due to inheritance and generic).

> @dsyme: I'm inclined not to do "type inference on record field names" for C# record types.

This may be a separate suggestion, and of course, mandated by F# adopting the initializer syntax and discouraging the positional constructor argument one.

What, overall, makes it important / better to keep distinction?

It feels with what @dsyme proposes, it would make consumption of F# and C# defined records in C# natural, and make consumption of records not defined in F# a bit alien in F#, with more "object" feel than seems required to me.

Just trying to understand better the motives, and having better code gen for F# records to be consumed in C# is nice move.

**Comment by smoothdeveloper** [comment]
Another point to consider is pattern matching, which has special syntax support in F#.

Related: #968

**Comment by Happypig375** [comment]
Record creation syntax is tricky. It is easy to see the simple case (one primary or parameterless constructor + all required fields) and think of the F# record creation syntax for it.

But how about
- Optional parameters in positional records?
- Other optionally initable properties?
- Additional constructors? You can't determine the primary constructor by attributes.
- Inference of record type from field names given all the optionality above?

If your answer is "record creation syntax can only applied with known type with parameters from all possible constructors and all possible properties", then why not just use the type constructor syntax that explicitly overloads on the record type?

It gets complex fast, you can't use the simple record syntax view to look at C# records. It is best to use constructor + properties to represent correct semantics instead of trying to modify record creation syntax.

Record update syntax should be fine though, and record deconstruction should just be object patterns (https://github.com/fsharp/fslang-suggestions/issues/968) indeed.

**Comment by xperiandri** [comment]
Let's support only the simplest case – default constructor

**Comment by Happypig375** [comment]
@xperiandri Will there be an error if there is more than one constructor? How should tooling (IntelliSense, compiler, Go To Definition...) indicate a record that's usable with `{ }` record creation syntax as opposed to a record that cannot? How would optional parameters interact with this, they are still possible in records with one primary constructor and no required fields?

How about record structs? Unlike record classes, record structs have no `<Clone>$` member. Would we make all structs usable with the record creation syntax? C# allowed [`with` syntax on _all_ structs](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-10.0/record-structs.md#allow-with-expression-on-structs).

**Comment by T-Gro** [comment]
Reviewed as part of RFC FS-1150 analysis: https://github.com/fsharp/fslang-design/pull/800#issuecomment-3852354596

   **Approved:**
   ```fsharp
   // Copy-and-update for C# records
   let updated = { csharpRecord with Name = "new" }
   ```

  **Not covered in this review:**, needs to be detailed out in an RFC:
   - Type inference on C# record field names, esp. at construction time

  Recommend a focused RFC that covers the missing bits apart from copy-and-update.
